enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/transaction/log/checkpoint/ConfigurableCheckPointFlushControl.java

```diff
@@ -0,0 +1,122 @@
+{
+    private static final AtomicIntegerFieldUpdater<ConfigurableCheckPointFlushControl> rushUpdater =
+            AtomicIntegerFieldUpdater.newUpdater( ConfigurableCheckPointFlushControl.class, "rushCount" );
+
+    private static final int NO_LIMIT = 0;
+    private static final int QUANTUM_MILLIS = 100;
+    private static final int TIME_BITS = 32;
+    private static final long TIME_MASK = (1L << TIME_BITS) - 1;
+
+    private final int iopq; // IOs per quantum
+    private final ObjLongConsumer<Object> pauseNanos;
+    private final Rush rush;
+
+    @SuppressWarnings( "unused" ) // Updated via rushUpdater
+    private volatile int rushCount;
+
+    public ConfigurableCheckPointFlushControl( Config config )
+    {
+        this( config, LockSupport::parkNanos );
+    }
+
+    // Only visible for testing
+    ConfigurableCheckPointFlushControl( Config config, ObjLongConsumer<Object> pauseNanos )
+    {
+        this.pauseNanos = pauseNanos;
+        int quantumsPerSecond = (int) (TimeUnit.SECONDS.toMillis( 1 ) / QUANTUM_MILLIS);
+        Integer iops = config.get( GraphDatabaseSettings.check_point_iops_limit );
+        this.iopq = iops == null || iops < 1? NO_LIMIT : (iops / quantumsPerSecond);
+        rush = () -> rushUpdater.getAndDecrement( this );
+    }
+
+    @Override
+    public IOLimiter getIOLimiter()
+    {
+
+        return iopq == NO_LIMIT? IOLimiter.unlimited() : this;
+    }
+
+    @Override
+    public Rush beginTemporaryRush()
+    {
+        rushUpdater.getAndIncrement( this );
+        return rush;
+    }
+
+    // The stamp is in two 32-bit parts:
+    // The high bits are the number of IOs performed since the last pause.
+    // The low bits is the 32-bit timestamp in milliseconds (~25 day range) since the last pause.
+    // We keep adding summing up the IOs until either a quantum elapses, or we've exhausted the IOs we're allowed in
+    // this quantum. If we've exhausted our IOs, we pause for the rest of the quantum.
+    // We don't make use of the Flushable at this point, because IOs from fsyncs have a high priority, so they
+    // might jump the IO queue and cause delays for transaction log IOs. Further, fsync on some file systems also
+    // flush the entire IO queue, which can cause delays on IO rate limited cloud machines.
+    // We need the Flushable to be implemented in terms of sync_file_range before we can make use of it.
+
+    @Override
+    public long maybeLimitIO( long previousStamp, int recentlyCompletedIOs, Flushable flushable ) throws IOException
+    {
+        if ( rushCount > 0 )
+        {
+            return INITIAL_STAMP;
+        }
+
+        long now = currentTimeMillis() & TIME_MASK;
+        long then = previousStamp & TIME_MASK;
+
+        if ( now - then > QUANTUM_MILLIS )
+        {
+            return now + (((long) recentlyCompletedIOs) << TIME_BITS);
+        }
+
+        long ioSum = (previousStamp >> TIME_BITS) + recentlyCompletedIOs;
```

Comment:
I don't think it's going to be a problem in practice. We have 32 bits for this count, so that's about 4.3 billion. Even `memcpy` can't go that fast. We actually have to overflow `long` to mess up the `ioSum >= iopq`, which is not possible.